source/MaterialXCore/Element.cpp

```diff
@@ -81,6 +81,84 @@ bool Element::operator!=(const Element& rhs) const
     return !(*this == rhs);
 }
 
+bool Element::isEquivalent(ConstElementPtr rhs, ElementEquivalenceOptions& options,
+                           ElementEquivalenceResult* result) const
+{
+    if (getCategory() != rhs->getCategory() ||
+        getName() != rhs->getName())
+    {
+        if (result)
+            result->addMessage("Name or category does not match for elements: '" + getNamePath() + "', '" + rhs->getNamePath() + "'");
+        return false;
+    }
+
+    // Compare attribute names. 
+    StringVec attributeNames = getAttributeNames();
+    StringVec rhsAttributeNames = rhs->getAttributeNames();
+
+    // Filter out any attributes specified in the options.
+    const StringSet& skipAttributes = options.skipAttributes;
+    if (!skipAttributes.empty())
+    {
+        attributeNames.erase(std::remove_if(attributeNames.begin(), attributeNames.end(),
+            [&skipAttributes](const string& attr) { return skipAttributes.find(attr) != skipAttributes.end(); }),
+            attributeNames.end());
+        rhsAttributeNames.erase(std::remove_if(rhsAttributeNames.begin(), rhsAttributeNames.end(),
+            [&skipAttributes](const string& attr) { return skipAttributes.find(attr) != skipAttributes.end(); }),
+            rhsAttributeNames.end());
+    }    
+
+    // Ignore ordering if option specified.
+    if (options.ignoreAttributeOrder)
+    {
+        // Sort the string arrays
+        std::sort(attributeNames.begin(), attributeNames.end());
+        std::sort(rhsAttributeNames.begin(), rhsAttributeNames.end());
+    }
+    if (attributeNames != rhsAttributeNames)
+    {
+        if (result)
+            result->addMessage("Attribute names differ for elements: '" + getNamePath() + "', '" + rhs->getNamePath() + "'");
```

Comment:
Just in the interest of keeping this feature as simple as possible, I'm curious about the situation in which these messages would be useful to a developer or an artist.  Would these messages be displayed in some fashion, allowing a developer or artist to respond to the differences that were found?

Reply:
That is my thought at the moment -- in the similar vein of `validate()` also providing feedback on where validation fails this is where equivalence has failed, can be corrected and the test re-run again. This is what I found frustrating with `operator==` where I had to modify the code to find out what differed.